### PR TITLE
Upgrade rust toolchain to nightly-2023-06-20

### LINKED ISSUE #2544: Update rust toolchain version to nightly-2023-06-20
Update the rust toolchain used by Kani.

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

Change Kani's compiler and tests to adapt to changes done to the toolchain. The biggest changes were:
 1. Implement overflow checks as part of `rvalue.rs` instead of intrinsics due to lowering of intrinsics to `BinOp`.
 2. Add a unsupported check for handling C string literals.

Here is a list of changes that I found that impacted this upgrade:

- https://github.com/rust-lang/rust/pull/108801
- https://github.com/rust-lang/rust/pull/112168
- https://github.com/rust-lang/rust/pull/112238
- https://github.com/rust-lang/rust/pull/112638
- https://github.com/rust-lang/rust/pull/110806
- https://github.com/rust-lang/rust/pull/112006
- https://github.com/rust-lang/rust/pull/111677
- https://github.com/rust-lang/rust/pull/111626
- https://github.com/rust-lang/rust/pull/109953
- https://github.com/rust-lang/rust/pull/112060
- https://github.com/rust-lang/rust/pull/112075
- https://github.com/rust-lang/rust/pull/111625
- https://github.com/rust-lang/rust/pull/111082
- https://github.com/rust-lang/rust/pull/111633
- https://github.com/rust-lang/rust/pull/110040

### Resolved issues:

Resolves #2544 

### Related RFC:

Optional #ISSUE-NUMBER.

### Call-outs:

~~This is still a draft PR. Before publishing it, I still need to:~~
~~1. I'll pull a bunch of the tests changes to a separate PR before publishing it.~~
~~2. This is still BLOCKED by a performance degradation on one of our tests.~~ I created #2576 to track a performance issue caused by this upgrade. I modified the regression test to go around the performance issue and created a new test that still captures the issue and is part of the performance CI job instead.

Note: I don't think there is anything complicated to actually implement C string literal support, but decided to keep this simple for now. We would also likely want to add more thorough tests too.

### Testing:

* How is this change tested?

* Is this a refactor change?

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by celinval** [comment]
For the `itoa` test, I narrowed it down the performance issue to a performance regression when formatting an i8. Updating the test to a u8 makes it much faster. I am adding a new performance test and modifying the `itoa` test for now. I created #2576 to capture the performance regression.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 1]
```diff
@@ -341,6 +407,42 @@ impl<'tcx> GotocCtx<'tcx> {
         }
     }
 
+    /// Check that a division does not overflow.
+    /// For integer types, division by zero is UB, as is MIN / -1 for signed.
+    /// Note that the compiler already inserts these checks for regular division.
```
Comment (adpaco-aws) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:412]: I don't see any test that demonstrates this. Can we include the checks coming from `rustc` into them so we can verify they still happen?

Also, can we refer to an issue tracking the work of getting rid of either?

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 2]
```diff
@@ -341,6 +407,42 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        &mut self,
+        dividend: &Operand<'tcx>,
+        divisor: &Operand<'tcx>,
+        loc: Location,
+    ) -> Stmt {
+        let divisor_expr = self.codegen_operand(divisor);
+        let div_by_zero_check = self.codegen_assert_assume(
+            divisor_expr.clone().is_zero().not(),
+            PropertyClass::ArithmeticOverflow,
+            "Attempt to divide by 0",
```
Comment (adpaco-aws) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:425]:
```suggestion
            "attempt to divide by 0",
```

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 3]
```diff
@@ -341,6 +407,42 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        if self.operand_ty(dividend).is_signed() {
+            let dividend_expr = self.codegen_operand(dividend);
+            let overflow_expr = dividend_expr
+                .clone()
+                .eq(dividend_expr.typ().min_int_expr(self.symbol_table.machine_model()))
+                .and(divisor_expr.clone().eq(Expr::int_constant(-1, divisor_expr.typ().clone())));
+            let overflow_check = self.codegen_assert_assume(
+                overflow_expr.not(),
+                PropertyClass::ArithmeticOverflow,
+                "Attempt to compute `MIN / -1` which would overflow",
```
Comment (adpaco-aws) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:437]:
```suggestion
                "attempt to compute `MIN / -1` which would overflow",
```

File: tests/perf/format/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This test checks the performance of calling format.
+//! This tests capture the performance regression introduced by the toolchain upgrade #2551.
+//! See https://github.com/model-checking/kani/issues/2576 for more details.
```
Comment (adpaco-aws) [inline tests/perf/format/src/lib.rs:6]:
```suggestion
//! See <https://github.com/model-checking/kani/issues/2576> for more details.
```

File: tests/perf/format/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This test checks the performance of calling format.
+//! This tests capture the performance regression introduced by the toolchain upgrade #2551.
```
Comment (adpaco-aws) [inline tests/perf/format/src/lib.rs:5]:
```suggestion
//! This test captures the performance regression introduced by the toolchain upgrade in <https://github.com/model-checking/kani/pull/2551>
```

File: tests/cargo-kani/itoa_dep/src/main.rs [hunk 6]
```diff
@@ -20,7 +20,7 @@ fn check_itoa<T: kani::Arbitrary + Integer + std::fmt::Display>() {
 #[kani::proof]
 #[kani::unwind(10)]
 fn check_signed() {
-    check_itoa::<i8>();
+    check_itoa::<u8>();
```
Comment (adpaco-aws) [inline tests/cargo-kani/itoa_dep/src/main.rs:23]: Can we instead comment out this harness (referring to #2576) and create another harness `check_unsigned` for the `u8` type?

File: kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs [hunk 7]
```diff
@@ -317,10 +317,12 @@ impl CodegenBackend for GotocCodegenBackend {
...
         _filenames: &OutputFilenames,
-    ) -> Result<(CodegenResults, FxHashMap<WorkProductId, WorkProduct>), ErrorGuaranteed> {
-        Ok(*ongoing_codegen
-            .downcast::<(CodegenResults, FxHashMap<WorkProductId, WorkProduct>)>()
-            .unwrap())
+    ) -> Result<(CodegenResults, FxIndexMap<WorkProductId, WorkProduct>), ErrorGuaranteed> {
+        match ongoing_codegen.downcast::<(CodegenResults, FxIndexMap<WorkProductId, WorkProduct>)>()
+        {
+            Ok(val) => Ok(*val),
+            Err(val) => panic!("Oh no: {:?}", val.type_id()),
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:324]: Oops. Let me fix that. LoL

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs (hunk 1 above)
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:412]: Yeah... Getting rid of them might be a bit tricky. But I can capture that in an issue if you prefer

Comment (adpaco-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:412]: Duplicate checks don't hurt for now but will complicate our post-processing moving forward. So yes, I'd like to capture it - the solution isn't necessarily to get rid of them, but long term we should agree on a way to uniformly report them and consider sanity checks to cover corner cases.

File: tests/cargo-kani/itoa_dep/src/main.rs (hunk 6 above)
Comment (celinval) [reply tests/cargo-kani/itoa_dep/src/main.rs:23]: Sure.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs [hunk 8]
```diff
@@ -227,6 +227,19 @@ impl<'tcx> GotocCtx<'tcx> {
                         );
                     }
                 }
+                ty::Adt(def, _) if Some(def.did()) == self.tcx.lang_items().c_str() => {
+                    // TODO: Handle CString
+                    // https://github.com/model-checking/kani/issues/2549
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs:232]: This is not a doc comment, so I don't think it makes a difference to add <>.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 9]
```diff
@@ -163,6 +163,52 @@ impl<'tcx> GotocCtx<'tcx> {
         }
     }
 
+    /// Generate code for a binary operation with an overflow and returns a tuple (res, overflow).
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:166]: The documentation was actually outdated, so thanks...

**Review by adpaco-aws — APPROVED** [review]
Thanks @celinval for this huge effort!

The situation with the duplicate checks is not ideal (will be confusing for users and may be downgrading performance), but we can improve it later.

**Comment by celinval** [comment]
> Thanks @celinval for this huge effort!
> 
> The situation with the duplicate checks is not ideal (will be confusing for users and may be downgrading performance), but we can improve it later.

Yeah, sorry about that. The problem already existed, and it just got even worse.  Fixing it will take a bit of time. Maybe I'll at least disable the redundant Cbmc checks for now, since that's an easy fix.
